onnxruntime/core/providers/cuda/controlflow/loop.cc

```diff
@@ -83,6 +82,21 @@ static Status ConcatenateGpuOutput(void* stream, std::vector<OrtValue>& per_iter
 }
 
 Loop::Loop(const OpKernelInfo& info) : onnxruntime::Loop(info) {
+  // We use the IDataTransfer abstraction to perform copies in the Loop implementation.
+  // By default, the GPUDataTransfer class is setup to use the same stream as the EP's compute stream
+  // while performing copies to/from CUDA (do_copy_on_default_stream = true). This is good as we wouldn't
+  // have to do any explicit syncs between the copy and compute streams.
+  // However, there is a user-facing flag that allows users to use a dedicated stream just for copying.
+  // To support using Loop for that case, we would have to do a sync between the copy stream and
+  // the compute stream to avoid data races. At the very least, we need to expose an interface in IDataTransfer
+  // to use a caller provided stream for Loop to provide for the GPUDataTransfer instance to use.
+  // Currently, using a dedicated copy stream has larger negative implications (see comment in GPUDataTransfer's
+  // constructor implementation), and so it is not in a useable state. When it becomes useable again,
+  // we will re-visit this limitation in Loop.
+  bool do_copy_on_default_stream = static_cast<const CUDAExecutionProvider*>(info.GetExecutionProvider())->DoCopyOnDefaultStream();
+  ORT_ENFORCE(do_copy_on_default_stream,
+              "Using Loop operator on CUDA while using a dedicated stream for copying "
+              "(a stream that is different than the compute stream) is currently not supported");
   SetConcatOutputFunc(ConcatenateGpuOutput);
```

Comment:
Is it guaranteed this value will not change after the kernels are created in session state finalization?

Reply:
Per my understanding, yes. They shouldn't be changed after the CUDA EP instance itself is created. These are values that flow through `OrtCudaProviderOptions` struct which reside in the constructed CUDA EP as `CUDAExecutionProviderInfo`...